src/handler/sign_tx.c

```diff
@@ -1,103 +1,423 @@
+        // cancel everything if user denies an output
+        explicit_bzero(&G_context, sizeof(G_context));
+        io_send_sw(SW_DENY);
+        return;
+    }
+    // decode next output and redisplay
+    switch (decode_next_element())
+    {
+    case TX_STATE_ERR:
+        io_send_sw(SW_INVALID_TX);
+        explicit_bzero(&G_context, sizeof(G_context));
+        ui_menu_main();
+        return;
+    case TX_STATE_PARTIAL:
+        // request more data with SW_OK
+        // THROW(SW_OK);
+        io_send_sw(SW_OK);
+        return;
+    case TX_STATE_READY:
+        // display new decoded_output
+        ui_display_tx_output(&confirm_tx_cb);
+        return;
+    case TX_STATE_FINISHED:
+        ui_display_tx_confirm();
+        return;
+    }
+}
 
-            return io_send_sw(SW_OK);
-        } else {  // last APDU, let's parse and sign
-            if (G_context.tx_info.raw_tx_len + cdata->size > MAX_TRANSACTION_LEN ||  //
-                !buffer_move(cdata,
-                             G_context.tx_info.raw_tx + G_context.tx_info.raw_tx_len,
-                             cdata->size)) {
-                return io_send_sw(SW_WRONG_TX_LENGTH);
-            }
+void init_sign_tx_ctx() {
+    explicit_bzero(&G_context, sizeof(G_context));
+    G_context.req_type = CONFIRM_TRANSACTION; /// SIGN_TX
 
-            G_context.tx_info.raw_tx_len += cdata->size;
+    G_context.tx_info.buffer_len = 0;
+    G_context.state = STATE_RECV_DATA;
+    G_context.tx_info.has_change_output = false;
+    G_context.tx_info.current_output = 0;
 
-            buffer_t buf = {.ptr = G_context.tx_info.raw_tx,
-                            .size = G_context.tx_info.raw_tx_len,
-                            .offset = 0};
+    cx_sha256_init(&G_context.tx_info.sha256);
+    G_context.tx_info.sighash_all[0] = '\0';
+}
+
+bool receive_data(buffer_t *cdata, uint8_t chunk) {
+    if (chunk == 0) {
+        if (G_context.state == STATE_RECV_DATA) {
+            // sent first chunk twice? return error
+            THROW(SW_BAD_STATE);
+        }
 
-            parser_status_e status = transaction_deserialize(&buf, &G_context.tx_info.transaction);
-            PRINTF("Parsing status: %d.\n", status);
-            if (status != PARSING_OK) {
-                return io_send_sw(SW_TX_PARSING_FAIL);
+        init_sign_tx_ctx();
+        // read change output and sighash
+        read_change_output_info(cdata);
+        // copy_sighash_all won't move cdata.offset
+        copy_sighash_all(cdata);
+        read_tx_data(cdata);
+
+        if(!buffer_move(cdata, G_context.tx_info.buffer, cdata->size - cdata->offset)) {
+            THROW(SW_WRONG_DATA_LENGTH);
+        }
+        G_context.tx_info.buffer_len = cdata->size - cdata->offset;
+        // return io_send_sw(SW_OK);
+        // check if user has already approved every needed info
+        // TODO ret: cannot receive more data after user approval
+    } else {
+        // copy sighash_all data to sha256 context
```

Comment:
Add comment explaining what is this case.

Reply:
Done.